File: src/src_platform_swing/fmg/swing/img/MosaicImg.java
package fmg.swing.img;

import java.util.Collection;

import fmg.core.img.MosaicAnimatedModel;
import fmg.core.mosaic.MosaicImageController;
import fmg.core.mosaic.cells.BaseCell;
import fmg.swing.mosaic.MosaicSwingView;

/**
 * Representable {@link fmg.core.types.EMosaic} as image
 * <br>
 * base SWING implementation
 *
 * @param <TImage> SWING specific image: {@link java.awt.Image} or {@link javax.swing.Icon}
 */
public abstract class MosaicImg<TImage>
        extends MosaicSwingView<TImage, Void, MosaicAnimatedModel<Void>>
{

    protected boolean _useBackgroundColor = true;

    protected MosaicImg() {
        super(new MosaicAnimatedModel<Void>());
    }

    @Override
    protected void drawBody() {
        //super.drawBody(); // !hide super implementation

        MosaicAnimatedModel<Void> model = getModel();

        _useBackgroundColor = true;
        switch (model.getRotateMode()) {
        case fullMatrix:
            drawModified(model.getMatrix());
            break;
        case someCells:
            // draw static part
            drawModified(model.getNotRotatedCells());

            // draw rotated part
            _useBackgroundColor = false;
            model.getRotatedCells(rotatedCells -> drawModified(rotatedCells));
            break;
        }
    }

    @Override
    public void close() {
        getModel().close();
        super.close();
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////
    //    custom implementations
    /////////////////////////////////////////////////////////////////////////////////////////////////////

    /** Mosaic image view implementation over {@link javax.swing.Icon} */
    static class Icon extends MosaicImg<javax.swing.Icon> {

        private IconSwing ico = new IconSwing(this);

        @Override
        protected javax.swing.Icon createImage() { return ico.create(); }

        @Override
        protected void drawModified(Collection<BaseCell> modifiedCells) {
            drawSwing(ico.getGraphics(), modifiedCells, _useBackgroundColor);
        }

        @Override
        public void close() {
            ico.close();
            super.close();
            ico = null;
        }

    }

    /** Mosaics image view implementation over {@link java.awt.Image} */
    static class Image extends MosaicImg<java.awt.Image> {

        private ImageAwt img = new ImageAwt(this);

        @Override
        protected java.awt.Image createImage() { return img.create(); }

        @Override
        protected void drawModified(Collection<BaseCell> modifiedCells) {
            img.drawWrapper(g -> drawSwing(g, modifiedCells, _useBackgroundColor));
        }

    }

    /** Mosaic image controller implementation for {@link Icon} */
    public static class ControllerIcon extends MosaicImageController<javax.swing.Icon, MosaicImg.Icon> {

        public ControllerIcon() {
            super(new MosaicImg.Icon());
        }

        @Override
        public void close() {
            getView().close();
            super.close();
        }

    }

    /** Mosaic image controller implementation for {@link Image} */
    public static class ControllerImage extends MosaicImageController<java.awt.Image, MosaicImg.Image> {

        public ControllerImage() {
            super(new MosaicImg.Image());
        }

        @Override
        public void close() {
            getView().close();
            super.close();
        }

    }

}

File: src/src_platform_uwp/fmg/uwp/img/win2d/MosaicImg.cs
using System.Collections.Generic;
using Windows.Graphics.Display;
using Windows.UI.Xaml;
using Microsoft.Graphics.Canvas;
using Microsoft.Graphics.Canvas.UI.Xaml;
using fmg.common;
using fmg.core.img;
using fmg.core.mosaic;
using fmg.core.mosaic.cells;
using fmg.uwp.utils;
using fmg.uwp.mosaic.win2d;

namespace fmg.uwp.img.win2d {

    /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image.
    /// <br/>
    /// Win2D implementation
    /// </summary>
    public static class MosaicImg {

        /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image. Base view Win2D implementation </summary>
        /// <typeparam name="TImage">Win2D specific image: <see cref="CanvasBitmap"/> or <see cref="CanvasImageSource"/></typeparam>
        public abstract class Win2DView<TImage> : MosaicWin2DView<TImage, Nothing, MosaicAnimatedModel<Nothing>>
            where TImage : DependencyObject, ICanvasResourceCreator
        {
            protected readonly ICanvasResourceCreator _rc;
            protected CanvasDrawingSession _ds;
            protected bool _useBackgroundColor = true;

            protected Win2DView(ICanvasResourceCreator resourceCreator)
                : base(new MosaicAnimatedModel<Nothing>())
            {
                _rc = resourceCreator;
            }

            protected override void DrawModified(ICollection<BaseCell> modifiedCells) {
                DrawWin2D(_ds, modifiedCells, _useBackgroundColor);
            }

            protected override void DrawBody() {
                //base.DrawBody(); // !hide base implementation

                MosaicAnimatedModel<Nothing> model = Model;

                _useBackgroundColor = true;
                switch (model.RotateMode) {
                case EMosaicRotateMode.fullMatrix:
                    DrawModified(model.Matrix);
                    break;
                case EMosaicRotateMode.someCells:
                    // draw static part
                    DrawModified(model.GetNotRotatedCells());

                    // draw rotated part
                    _useBackgroundColor = false;
                    model.GetRotatedCells(rotatedCells => DrawModified(rotatedCells));
                    break;
                }
            }

            protected override void Disposing() {
                base.Disposing();
                Model.Dispose();
            }

        }

        /////////////////////////////////////////////////////////////////////////////////////////////////////
        //    custom implementations
        /////////////////////////////////////////////////////////////////////////////////////////////////////

        /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image over <see cref="CanvasBitmap"/> </summary>
        public class CanvasBmpView : Win2DView<CanvasBitmap> {

            public CanvasBmpView(ICanvasResourceCreator resourceCreator)
                : base(resourceCreator)
            { }

            protected override CanvasBitmap CreateImage() {
                var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
                var s = Model.Size;
                return new CanvasRenderTarget(_rc, (float)s.Width, (float)s.Height, dpi);
            }

            protected override void DrawBody() {
                using (var ds = ((CanvasRenderTarget)Image).CreateDrawingSession()) {
                    _useBackgroundColor = true;
                    _ds = ds;
                    base.DrawBody();
                    _ds = null;
                }
            }

        }

        /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image over <see cref="CanvasImageSource"/> (XAML <see cref="Windows.UI.Xaml.Media.ImageSource"/> compatible) </summary>
        public class CanvasImgSrcView : Win2DView<CanvasImageSource> {

            public CanvasImgSrcView(ICanvasResourceCreator resourceCreator /* = CanvasDevice.GetSharedDevice() */)
                : base(resourceCreator)
            { }

            protected override CanvasImageSource CreateImage() {
                var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
                var s = Model.Size;
                return new CanvasImageSource(_rc, (float)s.Width, (float)s.Height, dpi);
            }

            protected override void DrawBody() {
                using (var ds = Image.CreateDrawingSession(Model.BackgroundColor.ToWinColor())) {
                    _useBackgroundColor = true;
                    _ds = ds;
                    base.DrawBody();
                    _ds = null;
                }
            }

        }

        /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image for <see cref="MosaicImg.CanvasBmpView"/> </summary>
        public class CanvasBmpController : MosaicImageController<CanvasBitmap, MosaicImg.CanvasBmpView> {

            public CanvasBmpController(ICanvasResourceCreator resourceCreator)
                : base(new MosaicImg.CanvasBmpView(resourceCreator))
            { }

            protected override void Disposing() {
                base.Disposing();
                View.Dispose();
            }

        }

        /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image for <see cref="MosaicImg.CanvasImgSrcView"/> </summary>
        public class CanvasImgSrcController : MosaicImageController<CanvasImageSource, MosaicImg.CanvasImgSrcView> {

            public CanvasImgSrcController(ICanvasResourceCreator resourceCreator)
                : base(new MosaicImg.CanvasImgSrcView(resourceCreator))
            { }

            protected override void Disposing() {
                base.Disposing();
                View.Dispose();
            }

        }

    }

}
